### PR TITLE
Add NPOT workaround for texture2d object.

### PR DESCRIPTION
**Pull request by 1chandu**
- Remove NPOT (non power of two) texture workaround from framebuffer object and add it to texure2d object. Make it applicable for only webgl1 context and when texture is NPOT.
- Fix assert failures in few lessons.
- Note: Merge is pending until deck.gl picking is verified.

### CODE REVIEW

File: src/utils/utils.js [hunk 1]
```diff
@@ -76,3 +78,13 @@ function detach(elem) {
 
   return ans;
 }
+
+/**
+ * Verifies if a given number is power of two or not.
+ * @param {object} n - The number to check.
+ * @return {Array} Returns true if the given number is power of 2, false otherwise.
+ **/
+export function isPowerOfTwo(n) {
```
Comment (ibgreen) [inline src/utils/utils.js]: Thought - if we don't need this in multiple places, then maybe put inside texture.js?

File: src/webgl/framebuffer.js [hunk 2]
```diff
@@ -187,15 +186,6 @@ export default class Framebuffer extends Resource {
         this._attachTexture({attachment, texture: object, layer: 0, level: 0});
       }
 
-      if (setFilteringParameters && object instanceof Texture2D) {
```
Comment (ibgreen) [inline src/webgl/framebuffer.js]: Two questions:
- Are there any requirements on filtering parameters in a texture when adding it to a framebuffer? I.e. could attachment fail if the filtering parameters are not correct? My assumption was that filtering has to be set to `GL.NEAREST`...

Note that framebuffer creates default attachments if none are specified. They need the right params.

File: src/webgl/texture.js [hunk 3]
```diff
@@ -210,6 +221,25 @@ export default class Texture extends Resource {
     };
     const glSettings = Object.assign({}, DEFAULT_TEXTURE_SETTINGS, pixelStore);
 
+    // NOTE: NPOT workaround: Force update default settings.
+    if (this._isNPOTWorkaroundApplicable()) {
+
+      log.warn(0, 'NPOT (non power of two) texture found, disabling mipmaping');
+      mipmaps = false;
+      if (parameters[this.gl.TEXTURE_MIN_FILTER] === undefined) {
```
Comment (ibgreen) [inline src/webgl/texture.js]: The `initialize()` function is a bit long, we should define helper functions and break it down.

File: src/webgl/texture.js [hunk 4]
```diff
@@ -680,12 +722,15 @@ export default class Texture extends Resource {
...
-      this.gl.bindTexture(this.target, this.handle);
       this.gl.texParameteri(this.target, pname, param);
       break;
     }
 
     this.gl.bindTexture(this.target, null);
     return this;
   }
+
+  _isNPOTWorkaroundApplicable() {
```
Comment (ibgreen) [inline src/webgl/texture.js]: Maybe a public function `isNPOT()` or `isPOT()` would be more useful?

File: src/webgl/texture.js [hunk 5]
```diff
@@ -666,6 +687,27 @@ export default class Texture extends Resource {
   _setParameter(pname, param) {
     this.gl.bindTexture(this.target, this.handle);
 
+    // NOTE: Apply NPOT workaround
```
Comment (ibgreen) [inline src/webgl/texture.js]: This looks like it could be broken out into a helper function.

File: src/webgl/texture.js [hunk 6]
```diff
@@ -210,6 +221,25 @@ export default class Texture extends Resource {
...
+      if (parameters[this.gl.TEXTURE_MIN_FILTER] === undefined) {
+        log.warn(0, 'NPOT (non power of two) texture found, forcing TEXTURE_MIN_FILTER to LINEAR');
+        parameters[this.gl.TEXTURE_MIN_FILTER] = this.gl.LINEAR;
+      }
+      if (parameters[this.gl.TEXTURE_WRAP_S] === undefined) {
+        log.warn(0, 'NPOT (non power of two) texture found, forcing TEXTURE_WRAP_S to CLAMP_TO_EDGE');
+        parameters[this.gl.TEXTURE_WRAP_S] = this.gl.CLAMP_TO_EDGE;
+      }
+      if (parameters[this.gl.TEXTURE_WRAP_T] === undefined) {
+        log.warn(0, 'NPOT (non power of two) texture found, forcing TEXTURE_WRAP_T to CLAMP_TO_EDGE');
```
Comment (ibgreen) [inline src/webgl/texture.js]: Use `log.once` to avoid error cascades.

File: src/webgl/texture.js [hunk 7]
```diff
@@ -666,6 +687,27 @@ export default class Texture extends Resource {
   _setParameter(pname, param) {
     this.gl.bindTexture(this.target, this.handle);
 
+    // NOTE: Apply NPOT workaround
+    if (this._isNPOTWorkaroundApplicable()) {
+      const supportedMinFilters = [GL.LINEAR, GL.NEAREST];
```
Comment (ibgreen) [inline src/webgl/texture.js]: Define a constant at top of file
`const NPOT_MIN_FILTERS`
you could make it an object, then you don't need to use indexOf.

File: src/utils/utils.js (hunk 1 above)
Comment (ibgreen) [inline src/utils/utils.js]: The util function `detach` above is it even used? If not, delete it.

File: test/utils/utils.spec.js [hunk 8]
```diff
@@ -30,3 +30,13 @@ test('Utils#uid', t => {
     'uid("prefix") starts with prefix');
   t.end();
 });
+
+test('Utils#isPowerOfTwo', t => {
```
Comment (ibgreen) [inline test/utils/utils.spec.js]: Nice!

File: test/webgl/texture.spec.js [hunk 9]
```diff
@@ -214,3 +214,66 @@ test('WebGL2#Texture2D setParameters', t => {
 
   t.end();
 });
+
+test('WebGL#Texture2D NPOT Workaround: texture creation', t => {
+  const {gl} = fixture;
```
Comment (ibgreen) [inline test/webgl/texture.spec.js]: Test cases should be run both on WebGL and WebGL2 context (only available in `test-browser`).

There should be some negative test cases, i.e. that test that the constructor throws exceptions when given NPOT textures under WebGL1.

File: src/webgl/framebuffer.js (hunk 2 above)
Comment (1chandu) [inline src/webgl/framebuffer.js]: I can't find any such filtering restrictions on the texture attachment, also I tested lesson-16 and picking on several deck-gl layers, all looking good.

Comment (ibgreen) [inline src/webgl/framebuffer.js]: OK - They should only play a role when rendering to textures. I have had problems with this in the past but perhaps unrelated.

Comment (ibgreen) [inline src/webgl/framebuffer.js]: Meant to say : should only play a role when **reading from** textures

File: src/utils/utils.js (hunk 1 above)
Comment (1chandu) [inline src/utils/utils.js]: `detach` is been used in `merge`, which is used in several other files.

File: src/webgl/framebuffer.js (hunk 2 above)
Comment (1chandu) [inline src/webgl/framebuffer.js]: Yeah, thats is what my understanding too, those settings should affect sampling only and should have no impact when writing to a render target.

File: src/webgl/texture.js (hunk 5 above)
Comment (1chandu) [inline src/webgl/texture.js]: done, moved to a helper method.

File: test/webgl/texture.spec.js (hunk 9 above)
Comment (1chandu) [inline test/webgl/texture.spec.js]: Wrote test cases for WebGL2, there are no exceptions to test, but I am testing parameter settings (for WebGL2, we shouldn't change these settings).

File: src/webgl/texture.js (hunk 5 above)
Comment (1chandu) [inline src/webgl/texture.js]: done, moved to a helper method.

File: test/setup/index.js [hunk 10]
```diff
@@ -14,7 +14,7 @@ export function createTestContext(opts = {}) {
 }
 
 export const fixture = {
-  gl: createTestContext(),
+  gl: createTestContext({webgl2: false, webgl1: true}),
```
Comment (1chandu) [inline test/setup/index.js]: createTestContext() with no input options using default options and hence creating WebGL2 context, I made this change to create WebGL1 context.

File: src/utils/utils.js (hunk 1 above)
Comment (1chandu) [inline src/utils/utils.js]: Keeping this function here, as it is pretty generic and nothing to do with texture object.

**Pull request merged**